### PR TITLE
feat(server): Allow configuration of hashtag extraction

### PR DESCRIPTION
**Pull request by chakaz**
Before this PR: Hashtags, if enabled, meant the text between `{` and `}` in the key (if exists and non-empty).

After this PR:

* Hashtags can _still_ be enabled / disabled
* Hashtag open / close char can be specified (and can be the same), like `:` popular with BullMQ
* Hashtag can include `N` closing tags to skip, like `{a}b}c}d` with `2` will return `a}b}c`.
* Hashtags can be limited to only be applied on keys with a specified prefix

This will allow some existing systems to migrate without code changes in client code.

### CODE REVIEW

File: src/server/common.cc [hunk 1]
```diff
@@ -29,33 +29,57 @@ extern "C" {
 ABSL_FLAG(bool, lock_on_hashtags, false,
           "When true, locks are done in the {hashtag} level instead of key level.");
 
+// We would have used `char` instead of `string` for the following 2 flags, but that's impossible.
+ABSL_FLAG(std::string, hashtag_open, "{", "Opening hashtag character. Must be a single char.");
+ABSL_FLAG(std::string, hashtag_close, "}", "Closing hashtag character. Must be a single char.");
+
+ABSL_FLAG(std::string, hashtag_prefix, "",
```
Comment (romange) [inline src/server/common.cc:36]: queue_prefix

Comment (romange) [reply src/server/common.cc:36]: locktag_prefix

File: src/server/common.cc [hunk 2]
```diff
@@ -29,33 +29,57 @@ extern "C" {
 ABSL_FLAG(bool, lock_on_hashtags, false,
           "When true, locks are done in the {hashtag} level instead of key level.");
 
+// We would have used `char` instead of `string` for the following 2 flags, but that's impossible.
+ABSL_FLAG(std::string, hashtag_open, "{", "Opening hashtag character. Must be a single char.");
```
Comment (romange) [inline src/server/common.cc:33]: locktag_delimiter

File: src/server/common.cc [hunk 3]
```diff
@@ -29,33 +29,57 @@ extern "C" {
...
+ABSL_FLAG(std::string, hashtag_close, "}", "Closing hashtag character. Must be a single char.");
+
+ABSL_FLAG(std::string, hashtag_prefix, "",
+          "Only keys with this prefix participate in hashtag extraction.");
+
+ABSL_FLAG(int, hashtag_close_skip_n_occurrence, 0,
```
Comment (romange) [inline src/server/common.cc:39]: unsigned and `locktag_skip_end` ?

File: src/server/common.cc [hunk 4]
```diff
@@ -26,36 +26,67 @@ extern "C" {
...
 
+// We would have used `char` instead of `string`, but that's impossible.
+ABSL_FLAG(
+    std::string, locktag_delimiter, "",
+    "If set, this char is used to extract a lock key by looking at delimiters, like hash tags. If "
+    "unset, regular hashtag extraction is done. Must be used with --lock_on_hashtags");
+
+ABSL_FLAG(unsigned, locktag_skip_n_end_delimiters, 0,
+          "How many closing hashtags should we skip. 0 for no skipping. For example, when set to "
+          "2, the hashtag for '{a}b}c}d}e' will be 'a}b}c'.");
```
Comment (romange) [inline src/server/common.cc:42]: I suggest changing the example to ':' delimiters - it's more clear and relevant, imho.

File: src/server/common.h [hunk 5]
```diff
@@ -53,8 +53,16 @@ class CommandId;
...
+  bool enabled = false;
+  char open_locktag = '{';
+  char close_locktag = '}';
+  unsigned skip_n_end_delimiters = 0;
+  std::string prefix;
+};
+
 struct KeyLockArgs {
-  static bool IsLockHashTagEnabled();
+  static LockTagOptions GetLockTagOptions();
```
Comment (romange) [inline src/server/common.h:65]: I suggest keeping `IsLockHashTagEnabled` and move `LockTagOptions` to cc

**Review by romange — APPROVED** [review]

File: src/server/common.h (hunk 5 above)
Comment (chakaz) [reply src/server/common.h:65]: But we need to expose these options for cluster_config, where hash tag extraction is done.
We can move it back, and maybe rename the `KeyTag()` function while on it, but it's kinda beyond the scope here I think.

**Review by romange — APPROVED** [review]

**Pull request merged**